trl/trainer/kto_trainer.py

```diff
@@ -0,0 +1,1109 @@
+                    ).logits
+
+        return self.get_batch_logps(
+            all_logits,
+            padded_batch["completion_labels"],
+            average_log_prob=False,
+            is_encoder_decoder=self.is_encoder_decoder,
+            label_pad_token_id=self.label_pad_token_id,
+        )
+
+    def build_tokenized_answer(self, prompt, answer):
+        """
+        Llama tokenizer does not satisfy `enc(a + b) = enc(a) + enc(b)`.
+        It does ensure `enc(a + b) = enc(a) + enc(a + b)[len(enc(a)):]`.
+        Reference:
+            https://github.com/EleutherAI/lm-evaluation-harness/pull/531#issuecomment-1595586257
+        """
+
+        full_tokenized = self.tokenizer(prompt + answer, add_special_tokens=False)
+        prompt_input_ids = self.tokenizer(prompt, add_special_tokens=False)["input_ids"]
+
+        answer_input_ids = full_tokenized["input_ids"][len(prompt_input_ids) :]
+        answer_attention_mask = full_tokenized["attention_mask"][len(prompt_input_ids) :]
+
+        # Concat tokens to form `enc(a) + enc(a + b)[len(enc(a)):]`
+        full_concat_input_ids = np.concatenate([prompt_input_ids, answer_input_ids])
+
+        # Prepare input tokens for token by token comparison
+        full_input_ids = np.array(full_tokenized["input_ids"])
+
+        if len(full_input_ids) != len(full_concat_input_ids):
+            raise ValueError("Prompt input ids and answer input ids should have the same length.")
+
+        # On some tokenizers, like Llama-2 tokenizer, there are occasions where tokens
+        # can be merged together when tokenizing prompt+answer. This could result
+        # on the last token from the prompt being different when tokenized on its own
+        # vs when done as prompt+answer.
+        response_token_ids_start_idx = len(prompt_input_ids)
+
+        # If tokenized prompt is different than both prompt+answer, then it means the
+        # last token has changed due to merging.
+        if prompt_input_ids != full_tokenized["input_ids"][:response_token_ids_start_idx]:
+            response_token_ids_start_idx -= 1
+
+        prompt_input_ids = full_tokenized["input_ids"][:response_token_ids_start_idx]
+        prompt_attention_mask = full_tokenized["attention_mask"][:response_token_ids_start_idx]
+
+        if len(prompt_input_ids) != len(prompt_attention_mask):
+            raise ValueError("Prompt input ids and attention mask should have the same length.")
+
+        answer_input_ids = full_tokenized["input_ids"][response_token_ids_start_idx:]
+        answer_attention_mask = full_tokenized["attention_mask"][response_token_ids_start_idx:]
+
+        return dict(
+            prompt_input_ids=prompt_input_ids,
+            prompt_attention_mask=prompt_attention_mask,
+            input_ids=answer_input_ids,
+            attention_mask=answer_attention_mask,
+        )
+
+    def shuffle_completion(self, batch) -> Dict:
```

Comment:
I found an issue with this function where if your dataset has extra columns besides "prompt", "completion", and "label", the `datasets` library will error due to the columns having inconsistent lengths (kl, prompt, etc will have lengths 2x the extra columns). Could we just double the lengths of all columns in case there are extras?

Reply:
@dblakely  can you kindly test now and report back?

Reply:
Hey @kashif, I got the same error about the columns having mismatched lengths:

```
pyarrow.lib.ArrowInvalid: Column 18 named prompt expected length 128 but got length 256
```

Not sure why because the way you did it looks like it should work to me.

However, maybe another way is to drop the columns except for prompt, completion, and label columns. For example:

```python
columns_to_remove = [col for col in train_dataset.column_names if col not in ["prompt", "completion", "label"]]
train_dataset = train_dataset.map(
    self.shuffle_completion, batched=True, batch_size=128, remove_columns=columns_to_remove
)
```

(and the `shuffle_completion` method would be reverted to how you originally had it)

Reply:
ok sure let's do that then...

Reply:
done